plugins/wporg-5ftf/includes/email.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+/**
+ * Helper functions for sending emails, including authentication tokens.
+ *
+ * We don't want pledges connected to individual w.org accounts, because that would encourage people to create
+ * "company" accounts instead of having their contributions show up as real human beings; or because that
+ * individual will likely eventually leave the company, and "ownership" of the pledge would be orphaned; or
+ * because we'd have to tie multiple accounts to the pledge (and all the extra time/UX costs associated with that),
+ * and would still have problems with orphaned ownership, etc.
+ *
+ * So instead, we just ask companies to create pledges using a group email (e.g., support@wordcamp.org), and
+ * we email them time-restricted, once-time-use auth tokens when they want to "log in".
+ *
+ * WP "nonces" aren't ideal for this purpose from a security perspective, because they're less secure. They're
+ * reusable, last up to 24 hours, and have a much smaller search space in brute force attacks. They also create an
+ * inconsistent UX, because a token could be valid for 24 hours, or for 1 second, due to how `wp_nonce_tick()`
+ * works. That would lead to some situations where a nonce had already expired by the time the contributor opened
+ * the email and clicked on the link.
+ *
+ * So instead, true NONCEs are implemented; see `is_valid_authentication_token()` for details.
+ */
+
+namespace WordPressDotOrg\FiveForTheFuture\Email;
+
+defined( 'WPINC' ) || die();
+
+const TOKEN_PREFIX = '5ftf_auth_token_';
+
+/**
+ * Wrap `wp_mail()` with shared functionality.
+ *
+ * @param string $to
+ * @param string $subject
+ * @param string $message
+ *
+ * @return bool
+ */
+function send_email( $to, $subject, $message ) {
+	$headers = array(
+		'From: WordPress - Five for the Future <donotreply@wordpress.org>',
+		'Reply-To: support@wordcamp.org',
+			// todo update address when new one is created
+	);
+
+	return wp_mail( $to, $subject, $message, $headers );
+}
+
+/**
+ * Generate an action URL with a unique authentication token.
+ *
+ * @param int    $pledge_id
+ * @param string $action
+ * @param int    $action_page_id The ID of the page that the user will be taken back to, in order to process their
+ *                               verification request.
+ *
+ * @return string
+ */
+function get_authentication_url( $pledge_id, $action, $action_page_id ) {
+	$auth_token = array(
+		'value'      => wp_generate_password( 20, false ), // Similar to `get_password_reset_key()`.
+			// todo should encrypt at rest? core doesn't but others do
+		'expiration' => time() + ( 2 * HOUR_IN_SECONDS ),
+	);
```

Comment:
🤔 

I think the existing setup will work for contributor emails, since they're basically the same flow.

That's a good point about managing pledges, though. My previous thoughts were to still use this, but maybe add context to it to control that kind of thing; related discussion: https://github.com/WordPress/five-for-the-future/pull/46/files#r339103839. 

I played around w/ modularizing it -- which is normally something I've very much in favor of -- but got a bit nervous about human error and mishandling the token. We're also on a tight deadline, and don't know what our future needs will look like, so my vote would be for leaving it as-is, and then reconsidering refactoring it when we have a tangible use case.

Reply:
👍